src/cargo/util/command_prelude.rs

```diff
@@ -986,6 +989,49 @@ pub fn root_manifest(manifest_path: Option<&Path>, gctx: &GlobalContext) -> Carg
     }
 }
 
+pub fn get_targets_from_rustup() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustup")
+        .arg("target")
+        .arg("list")
+        .arg("--installed")
+        .output()?;
+
+    if output.status.success() {
+        let stdout = String::from_utf8(output.stdout)?;
+        return Ok(stdout.lines().map(|line| line.to_string()).collect());
+    }
+
+    Ok(vec![])
+}
+
+pub fn get_target_from_rustc() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustc")
+        .arg("--print")
+        .arg("target-list")
+        .output()?;
+
+    if output.status.success() {
+        let stdout = String::from_utf8(output.stdout)?;
+        return Ok(stdout.lines().map(|line| line.to_string()).collect());
+    }
+
+    Ok(vec![])
```

Comment:
nit: I generally recommend early guard statements.  The structure of the code (blocks, indentations) should match the business logic.  The way this is written, it makes the success case look like its either not going to return or, if it does, that it will, that its an error case

Contrast it with
```suggestion
    let output = std::process::Command::new("rustc")
        .arg("--print")
        .arg("target-list")
        .output()?;
    if !output.status.success() {
        return Ok(vec![]);
    }

    let stdout = String::from_utf8(output.stdout)?;
    Ok(stdout.lines().map(|line| line.to_owned()).collect())
```
(I also recommend `to_owned` over `to_string` as `to_string` works with any `Display` type and can lead to bugs with refactors)